File: backend/app/services/valuation.py
"""Wallet valuation helpers shared by leaderboards, portfolios and snapshots."""

import logging
from decimal import Decimal

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.models import Membership, PriceSnapshot
from app.services import market

logger = logging.getLogger(__name__)

TWO_PLACES = Decimal("0.01")


def get_price_with_fallback(db: Session, ticker: str) -> Decimal | None:
    """Live quote if possible, otherwise the most recent stored snapshot."""
    try:
        return market.get_quote(ticker)
    except market.UnknownTickerError:
        pass
    except Exception:
        logger.warning("live quote failed for %s, using snapshot", ticker, exc_info=True)

    snapshot = db.scalars(
        select(PriceSnapshot)
        .where(PriceSnapshot.ticker == ticker)
        .order_by(PriceSnapshot.snapshot_date.desc())
        .limit(1)
    ).first()
    return snapshot.price if snapshot else None


def get_holdings_value(db: Session, membership: Membership) -> tuple[Decimal, list[dict]]:
    """Returns (total position value, per-holding breakdown)."""
    total = Decimal("0")
    breakdown = []
    for holding in membership.holdings:
        price = get_price_with_fallback(db, holding.ticker)
        value = (holding.shares * price).quantize(TWO_PLACES) if price is not None else None
        if value is not None:
            total += value
        breakdown.append(
            {
                "ticker": holding.ticker,
                "shares": holding.shares,
                "avg_cost": holding.avg_cost,
                "current_price": price,
                "market_value": value,
                "cost_basis": (holding.shares * holding.avg_cost).quantize(TWO_PLACES),
            }
        )
    return total, breakdown


def get_total_value(db: Session, membership: Membership) -> Decimal:
    holdings_value, _ = get_holdings_value(db, membership)
    return (membership.cash_balance + holdings_value).quantize(TWO_PLACES)

File: backend/src/valuation.js
import { getQuote, UnknownTickerError } from "./market.js";
import { r2 } from "./util.js";

/** Live quote if possible, otherwise the most recent stored daily snapshot. */
export async function getPriceWithFallback(env, ticker) {
  try {
    return await getQuote(ticker);
  } catch (error) {
    if (!(error instanceof UnknownTickerError)) console.warn("quote failed", ticker, error);
  }
  const snapshot = await env.DB.prepare(
    "SELECT price FROM price_snapshots WHERE ticker = ? ORDER BY snapshot_date DESC LIMIT 1"
  ).bind(ticker).first();
  return snapshot?.price ?? null;
}

/** Values a set of holding rows (live wallets). Returns {total, breakdown}. */
export async function valueHoldings(env, holdingRows) {
  let total = 0;
  const breakdown = [];
  for (const holding of holdingRows) {
    const price = await getPriceWithFallback(env, holding.ticker);
    const value = price === null ? null : r2(holding.shares * price);
    if (value !== null) total += value;
    breakdown.push({
      ticker: holding.ticker,
      shares: holding.shares,
      avg_cost: holding.avg_cost,
      current_price: price,
      market_value: value,
      cost_basis: r2(holding.shares * holding.avg_cost),
    });
  }
  return { total: r2(total), breakdown };
}

export async function membershipHoldings(env, membershipId) {
  const { results } = await env.DB.prepare(
    "SELECT * FROM holdings WHERE membership_id = ? ORDER BY ticker"
  ).bind(membershipId).all();
  return results;
}
